torch/csrc/jit/tensorexpr/half_support.h

```diff
@@ -101,7 +101,8 @@ class HalfRewriter : public IRMutator {
     // just don't allow half casts we didn't insert.
     if (isHalf(v)) {
       if (inserted_half_casts_.count(v) < 1) {
-        return child;
+        v->set_dtype(v->dtype().cloneWithScalarType(c10::kFloat));
+        return v;
```

Comment:
Shouldn't this be `child` instead of `v`? And why do we need this change at all?

Reply:
The child value could be long or double as follows. Then the immediate long value `2` needs to be converted to bloat16. But we do not support the conversion between non-float32 and bfloat16. Hence, I always update the `Cast` to cast the non-float32 to float32.
```python
a = torch.randn(10).bfloat16()
c = a + 2
```

Reply:
I think in that case we need to insert an extra cast then? I.e. first convert to f32, then from f32 to bf16. I'm not sure bluntly overwriting dtype is a correct solution here. Also, could you please add tests exercising these type conversions? I believe current tests might not be covering that.

Reply:
> I think in that case we need to insert an extra cast then? I.e. first convert to f32, then from f32 to bf16.

It is sort of complex. Because the immediate value of the python script will be translated to `LongImm` or `DoubleImm`, and it is hard to know if we need to insert `Cast` because it depends on the `Expression`. Take an example as follows, the `2` need to be cast to `Float` but not `BFloat16`.

```python
a = torch.randn(10)
c = a + 2
```

Reply:
Just clarify, you will cast non-float32 to float32 first. And then cast to bfloat16 or no change depends on current expression? If we add cast there, the purpose is also casting it to float32. I feel there is no different there.

Reply:
> you will cast non-float32 to float32 first. And then cast to bfloat16 or no change depends on current expression?

Yes, we always cast the non-float32 to float32 first and then cast to bfloat16 if needed. 

> If we add cast there, the purpose is also casting it to float32. I feel there is no difference there.

If we add cast for each immediate data type, we have to enumerate all the Immediate value types like `LongImm`, `IntImm`, `ByteImm`. So that's why I just change the destination datatype of `cast` to `float`

Reply:
> If we add cast for each immediate data type, we have to enumerate all the Immediate value types like LongImm, IntImm, ByteImm. So that's why I just change the destination datatype of cast to float

If that's what needs to be done to accurately represent the original behavior, then we should do that. Otherwise we will simply break models when these casts actually happen.

Reply:
> If that's what needs to be done to accurately represent the original behavior, then we should do that. Otherwise we will simply break models when these casts actually happen.

I'm not sure if the immediate value is the only case. But anyway, I will double-check it.